flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedLeaderService.java

```diff
@@ -356,8 +357,8 @@ public void confirmLeaderSessionID(UUID leaderSessionID) {
 		}
 
 		@Override
-		public boolean hasLeadership() {
-			return isLeader;
+		public boolean hasLeadership(@Nonnull UUID leaderSessionId) {
+			return isLeader && leaderSessionId.equals(currentLeaderSessionId);
 		}
```

Comment:
I would ask, in what situation, the ```leaderSessionId``` does not equal to ```currentLeaderSessionId``` . When another task update the value ?

Reply:
In the case of the `EmbeddedLeaderService`, there is no case in which `leaderSessionId != currentLeaderSessionId` if `isLeader`, because a leader only loses leadership if the service is stopped. But this is an implementation detail which might change in the future and thus, should be guarded.

I think we don't need a lock but have to make `currentLeaderSessionId` `volatile` such that we see modifications.

Reply:
Thanks @tillrohrmann Makes sense better to me now. Yea, I did not see there has a volatile in there before. Thanks again.